### ISSUE TITLE
allow zero-copy/no-serialization services for intra-process communication

### ISSUE THREAD

**Issue by protobits** [ros/ros_comm#514]
I'm using nodelets and I believe service call messages are being serialized even when conforming to the intra-process communication pattern. I've found out about it after adapting a custom C++ class to a ROS message using message traits. The serialization method for this class does nothing since this message is only to be sent between nodes and this works correctly (the serialization method is never called). However, defining Request/Response messages by adapting a custom C++ class in fact has the serialize method invoked.

If you provide some brief guidance about where this limitation is present in the code maybe I can provide some patch.

**Comment by dirk-thomas** [comment]
Yes, for services no intra-process communication is implemented.

Some pointers for you to look into it:
- the major part deciding about serializing or not for pub/sub happens in the `TopicManager::publish()` (https://github.com/ros/ros_comm/blob/2c4af852eafd5335dc2dff7446a1879519a8e2b6/clients/roscpp/src/libros/topic_manager.cpp#L705), something similar will be necessary for services
- the `ServiceServer` / `ServiceClient` classes are internally using `ServerServerLink` / `ServiceClientLink` instance which are specific for doing serialization and holding `Connection` instances, this area will likely need a different structure to also support zero-copy
- while pub / sub is inherently asynchronous services are currently providing a blocking interface, this will need to be considered when dispatching them through the single threaded or multi threaded call back queue of nodelets

For more details I would recommend to simply follow the call hierarchy and you will find several more areas involved in it.

I expect that implementing this feature will require a significant amount of changes and effort. Depending on how invasive a potential patch will be it will likely not be released into existing ROS distro but only be considered for future ROS distros (e.g. _Jade_).

**Comment by protobits** [comment]
Ok, seems a daunting task. I will have a look at it. If I get to understand on how to work it out I'll let you know.

**Comment by sloretz** [comment]
ROS 1 is end-of-life (EOL) as of today, May 31st 2025. I am archiving this repository because:

* it only supports ROS 1
* it isn't needed anymore in ROS 2

[If you still rely on ROS 1, read this page to learn about your options.](https://bit.ly/NoeticEOL)
